src/rowbinary/de.rs

```diff
@@ -242,42 +400,204 @@ impl<'data> Deserializer<'data> for &mut RowBinaryDeserializer<'_, 'data> {
             }
         }
 
-        visitor.visit_seq(Access {
-            deserializer: self,
+        let inner_data_type_validator = self.validator.validate(
+            &SerdeType::Tuple,
+            &[
+                DataTypeHint::Tuple,
+                DataTypeHint::Array,
+                DataTypeHint::FixedString,
+                // FIXME: uncomment when there is a way to implement ReverseSeqAccess
+                // DataTypeHint::IPv4,
+                DataTypeHint::IPv6,
+            ],
             len,
-        })
+        )?;
+        let mut new_self = RowBinaryDeserializer {
+            input: self.input,
+            validator: inner_data_type_validator,
+        };
+        let access = RowBinaryTupleAccess {
+            deserializer: &mut new_self,
+            len,
+        };
+        visitor.visit_seq(access)
     }
 
     #[inline]
     fn deserialize_option<V: Visitor<'data>>(self, visitor: V) -> Result<V::Value> {
-        ensure_size(&mut self.input, 1)?;
+        // println!("deserialize_option call");
 
+        ensure_size(&mut self.input, 1)?;
+        let inner_data_type_validator =
+            self.validator
+                .validate(&SerdeType::Option, &[DataTypeHint::Nullable], 0)?;
         match self.input.get_u8() {
-            0 => visitor.visit_some(&mut *self),
+            0 => visitor.visit_some(&mut RowBinaryDeserializer {
+                input: self.input,
+                validator: inner_data_type_validator,
+            }),
             1 => visitor.visit_none(),
```

Comment:
This is currently the main drawback of the validation implementation if we want to disable it after the first N rows for better performance. If these first rows are all NULLs, then we do not properly validate the inner type.